client/remoterepo.c

```diff
@@ -397,310 +482,51 @@ TDNFChecksumFromHexDigest(
-       !ml_file)
+       !ml_ctx)
     {
         dwError = ERROR_TDNF_INVALID_PARAMETER;
         BAIL_ON_TDNF_ERROR(dwError);
     }
 
-    metalink_context = metalink_parser_context_new();
-    if (metalink_context == NULL)
-    {
-        dwError = ERROR_TDNF_OUT_OF_MEMORY;
-        BAIL_ON_TDNF_ERROR(dwError);
-    }
-
     fd = open(pszFile, O_RDONLY);
     if (fd == -1)
     {
         dwError = errno;
         BAIL_ON_TDNF_SYSTEM_ERROR_UNCOND(dwError);
     }
-    while((length = read(fd, buf, (sizeof(buf)-1))) > 0)
-    {
-        metalink_error = metalink_parse_update(metalink_context, buf, length);
-        memset(buf, 0, BUFSIZ);
-        if (metalink_error != 0)
-        {
-            if (metalink_context)
-            {
-                metalink_parser_context_delete(metalink_context);
-            }
-            pr_err("Unable to parse metalink, ERROR: code=%d\n", metalink_error);
-            dwError = ERROR_TDNF_INVALID_PARAMETER;
-            BAIL_ON_TDNF_ERROR(dwError);
-        }
-    }
-    if (length == -1)
-    {
-        if (metalink_context)
-        {
-            metalink_parser_context_delete(metalink_context);
-        }
-        dwError = errno;
-        BAIL_ON_TDNF_SYSTEM_ERROR(dwError);
-    }
-    metalink_error = metalink_parse_final(metalink_context, NULL, 0, &metalink);
-    if ((metalink_error != 0) || (metalink == NULL))
-    {
-        pr_err("metalink_parse_final failed, ERROR: code=%d\n", metalink_error);
-        dwError = ERROR_TDNF_INVALID_PARAMETER;
-        BAIL_ON_TDNF_ERROR(dwError);
-    }
-    if (metalink->files == NULL)
-    {
-        pr_err("Metalink does not contain any valid file.\n");
-        dwError = ERROR_TDNF_INVALID_REPO_FILE;
-        BAIL_ON_TDNF_ERROR(dwError);
-    }
 
-    for(files = metalink->files; files && *files; ++files)
+    dwError = TDNFMetalinkParseFile(ml_ctx, fd, TDNF_REPO_METADATA_FILE_NAME);
+    if (dwError)
     {
-        resources = (*files)->resources;
-        if (IsNullOrEmptyString(resources))
-        {
-            pr_err("File %s does not have any resource.\n", (*files)->name);
-            dwError = ERROR_TDNF_METALINK_RESOURCE_VALIDATION_FAILED;
-            BAIL_ON_TDNF_ERROR(dwError);
-        }
-        while(*resources)
-        {
-            dwError = TDNFAllocateResourceURL(&urls_curr, (*resources)->url);
-            BAIL_ON_TDNF_ERROR(dwError);
-
-            if (!urls_head)
-            {
-                urls_head = urls_curr;
-            }
-            else
-            {
-                urls_prev->next = urls_curr;
-            }
-            urls_prev = urls_curr;
-            ++resources;
-        }
-
-        dwError = TDNFGetFileHashFromMetalink((*files), ml_file);
+        pr_err("Unable to parse metalink, ERROR: code=%d\n", dwError);
         BAIL_ON_TDNF_ERROR(dwError);
-        (*ml_file)->urls = urls_head;
     }
+ 
+    //sort the URL's in List based on preference.
+    TDNFSortListOnPreference(&(ml_ctx->urls));
```

Comment:
better to have a return value from this function. and no need to use braces for `(ml_ctx->urls)`